File: C++/maximalSquare.cpp
/**
 * @File    : maximalSquare.cpp
 * @Brief   : 最大正方形 动态规划
 * @Link    : https://leetcode-cn.com/problems/maximal-square/
 * @Author  : Wei Li
 * @Date    : 2021-07-14
*/

#include <iostream>
#include <vector>

class Solution
{
public:
    int maximalSquare(std::vector<std::vector<char>> &matrix)
    {
        if (matrix.empty() || matrix[0].empty())
        {
            return 0;
        }

        int m = matrix.size(), n = matrix[0].size(), max_side = 0;
        std::vector<std::vector<int>> dp(m + 1, std::vector<int>(n + 1, 0));
        for (int i = 1; i <= m; ++i)
        {
            for (int j = 1; j <= n; ++j)
            {
                if (matrix[i - 1][j - 1] == '1')
                {
                    dp[i][j] = std::min(dp[i - 1][j - 1], std::min(dp[i][j - 1], dp[i - 1][j])) + 1;
                }
                max_side = std::max(max_side, dp[i][j]);
            }
        }

        return max_side * max_side;
    }
};

// -------------------------------
int main(int argc, char **argv)
{
    std::vector<std::vector<char>> matrix = {{'1', '0', '1', '0', '0'}, 
                                            {'1', '0', '1', '1', '1'}, 
                                            {'1', '1', '1', '1', '1'}, 
                                            {'1', '0', '0', '1', '0'}};
    Solution solution;
    auto max_area = solution.maximalSquare(matrix);
    std::cout << "The solution of this problem is : " << max_area << std::endl;
    return 0;
}

File: Python/maximalSquare.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Filename: maximalSquare.py
@Function: 最大正方形 动态规划 
@Link: https://leetcode-cn.com/problems/maximal-square/
@Python Version: 3.8
@Author: Wei Li
@Date: 2021-07-14
"""

from typing import List


class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        if len(matrix) == 0 or len(matrix[0]) == 0:
            return 0

        maxSide = 0
        rows, columns = len(matrix), len(matrix[0])
        dp = [[0] * columns for _ in range(rows)]
        for i in range(rows):
            for j in range(columns):
                if matrix[i][j] == '1':
                    if i == 0 or j == 0:
                        dp[i][j] = 1
                    else:
                        dp[i][j] = min(dp[i - 1][j], dp[i]
                                       [j - 1], dp[i - 1][j - 1]) + 1
                    maxSide = max(maxSide, dp[i][j])

        maxSquare = maxSide * maxSide

        return maxSquare


# ---------------------------------
if __name__ == "__main__":
    matrix = [["1", "0", "1", "0", "0"], ["1", "0", "1", "1", "1"],
              ["1", "1", "1", "1", "1"], ["1", "0", "0", "1", "0"]]

    solution = Solution()
    max_area = solution.maximalSquare(matrix)
    print(f"The solution of thsi problem is {max_area}")
